vcf.c

```diff
@@ -1206,6 +1299,14 @@ bcf_hdr_t *bcf_hdr_init(const char *mode)
     if (!h) return NULL;
     for (i = 0; i < 3; ++i)
         if ((h->dict[i] = kh_init(vdict)) == NULL) goto fail;
+
+    bcf_hdr_aux_t *aux = (bcf_hdr_aux_t*)calloc(1,sizeof(bcf_hdr_aux_t));
+    if ( !aux ) goto fail;
+    if ( (aux->gen = kh_init(hdict))==NULL ) { free(aux); goto fail; }
+    aux->dict = *((vdict_t*)h->dict[0]);
+    free(h->dict[0]);
+    h->dict[0] = aux;
```

Comment:
We have:

```
    vdict_t dict;   // bcf_hdr_t.dict[0] vdict_t dictionary which keeps bcf_idinfo_t for BCF_HL_FLT,BCF_HL_INFO,BCF_HL_FMT
    hdict_t *gen;   // hdict_t dictionary which keeps bcf_hrec_t* pointers for generic and structured fields
} bcf_hdr_aux_t;
```

but this just causes me to scratch my head.  Why is one dictionary an actual struct and the other a pointer to one?  What's the significance?  Surely there has to be a meaning? 

Answer - no!  No meaning whatsoever, it's just a byproduct of a strange allocation policy policy that does a structure to structure copy.  Why is it like this?

IMO it would be saner with something like:

```
    vdict_t *dict;   // bcf_hdr_t.dict[0] vdict_t dictionary which keeps bcf_idinfo_t for BCF_HL_FLT,BCF_HL_INFO,BCF_HL_FMT
    hdict_t *gen;   // hdict_t dictionary which keeps bcf_hrec_t* pointers for generic and structured fields
} bcf_hdr_aux_t;
```

and for allocation purposes:

```
bcf_hdr_aux_t *aux = (bcf_hdr_aux_t*)calloc(1,sizeof(bcf_hdr_aux_t));
aux->dict = (vdict_t *)h->dict[0];
aux->gen = kh_init(hdict);
// if (!aux->gen) blah
```

Obviously the code that frees aux->gen would need updating to also destroy aux->dict and free the h->dict pointer as before, but this is all internal so that doesn't change ABI at all.

Reply:
Oh god... I see now!

My naive assumption that given hdr->dict is a void pointer and so is an opaque type that is not part of the public API, we do, sadly, have code that understands the private internals of htslib and manipulates it still. :-(

https://github.com/samtools/bcftools/blob/develop/vcfannotate.c#L279

So fair enough I see why it's like it is.  However it needs a comment documenting this.